core/src/main/java/com/google/cloud/sql/core/DefaultAccessTokenSupplier.java

```diff
@@ -0,0 +1,199 @@
+          credential.refreshToken();
+
+          return new AccessToken(
+              credential.getAccessToken(), getTokenExpirationTime(credential).orElse(null));
+        }
+      };
+    }
+    throw new RuntimeException(
+        String.format(
+            "Unable to connect via automatic IAM authentication: "
+                + "Unsupported credentials of type %s",
+            source.getClass().getName()));
+  }
+
+  /**
+   * Returns an access token, refreshing if the credentials are expired or nearly expired.
+   *
+   * @return the AccessToken or Optional.empty() if there is no tokenSource.
+   * @throws IOException if there is an error attempting to refresh the token
+   * @throws IllegalStateException if the token cannot be used to refresh.
+   */
+  @Override
+  public Optional<AccessToken> get() throws IOException {
+    if (tokenSource.isPresent()) {
+      RetryingCallable<Optional<AccessToken>> retries =
+          new RetryingCallable<>(
+              () -> {
+                final GoogleCredentials credentials;
+                credentials = parseCredentials();
+                try {
+                  credentials.refreshIfExpired();
+                } catch (IllegalStateException e) {
+                  throw new IllegalStateException("Error refreshing credentials " + credentials, e);
+                }
+                if (credentials.getAccessToken() == null
+                    || credentials.getAccessToken().equals("")) {
+                  throw new IllegalStateException("Credentials do not have an access token");
+                }
+                if (credentials.getAccessToken().getExpirationTime() != null
+                    && credentials.getAccessToken().getExpirationTime().before(new Date())) {
+                  throw new IllegalStateException(
+                      "Credentials were refreshed but expiration time is in the past");
+                }
+                GoogleCredentials downscoped = getDownscopedCredentials(credentials);
+                if (downscoped.getAccessToken() == null || downscoped.getAccessToken().equals("")) {
+                  throw new IllegalStateException(
+                      "Donwscoped credentials do not have an access token");
+                }
+                return Optional.of(downscoped.getAccessToken());
+              },
+              this.retryCount,
+              this.retryDuration);
+
+      try {
+        return retries.call();
+      } catch (IOException e) {
+        throw e;
+      } catch (RuntimeException e) {
+        throw e;
+      } catch (Exception e) {
+        throw new RuntimeException("Unexpected exception refreshing authentication token", e);
+      }
+    }
+    return Optional.empty();
+  }
+
+  /**
+   * Extracts the expiration time from an AccessToken.
+   *
+   * @param token the token
+   * @return the expiration time, if set.
+   */
+  static Optional<Date> getTokenExpirationTime(Optional<AccessToken> token) {
```

Comment:
Moved from [SqlAdminApiFetcher.java:161](https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory/blob/183c71ab5a6068b1f69d264cd72fbd1fb4fcbc17/core/src/main/java/com/google/cloud/sql/core/SqlAdminApiFetcher.java#L161)

Reply:
side note: we need to stop using Date and instead prefer Instance. If we don't do that as part of this PR, let's do it as part of the next one.

Reply:
Good idea. I think a separate PR is the way to go.

Reply:
Let's make sure to do this -- file an issue you're not planning on just getting to it right away.

Reply:
See #1340